### PR TITLE
Reduce snapshot from cpp sources

### PR DESCRIPTION
**Pull request by hhorii**
### Summary

#1591 removed snapshot instructions from python. This PR removes snapshots from cpp.

### Details and comments

Because of the removal of snapshot instructions in #1591, cpp implementation for them is no longer necessary.
This PR removes implementations to process snapshot instructions.

This PR will changes many sources. I think this should be back-ported to 0.11 to easily merge following PRs.

### CODE REVIEW

**Review by doichanj — CHANGES_REQUESTED** [review]
https://github.com/hhorii/qiskit-aer/blob/e7c66f66936c2a05d34c4b8251cad5df4812ae4b/src/simulators/statevector/statevector_state.hpp#L512-L516
there are snapshots in comment (also in other State classes) 
is JSON used for snapshots?

**Comment by doichanj** [comment]
https://github.com/hhorii/qiskit-aer/blob/e7c66f66936c2a05d34c4b8251cad5df4812ae4b/src/framework/operations.hpp#L238
I think this is not used for snapshot but used for save operations.

**Review by jakelishman — COMMENTED** [review]
This should not be backported - it's inherently not stable because it's a load of removals.  That said, I don't think it will be as big an issue for merging backports as you think, because it's just deletions, not changes.  The deleted code (presumably) hasn't been touched in a very long time, so it's unlikely it's going to be touched in the 0.11 series, especially now it's being deleted from 0.12.

File: src/framework/json.hpp [hunk 1]
```diff
@@ -121,6 +122,15 @@ void to_json(json_t &js, const std::vector<std::complex<RealType>> &vec);
...
 
+/**
+ * Convert a custom complex vector to a json list
+ * v -> [ [real(v[0]), imag(v[0])], ...]
+ * @param js a json_t object to contain converted type.
+ * @param vec a complex vector to convert.
+ */
+template <typename RealType>
+void to_json(json_t &js, const AER::Vector<std::complex<RealType>> &vec);
+
```
[the remark this answers is not in the document]
Comment (hhorii) [reply src/framework/json.hpp:133]: This code was necessary to build the codes. Error messages are as follows.
```
~/.conan/data/nlohmann_json/3.1.1/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/nlohmann/detail/conversions/to_json.hpp:312:9: error: static_assert failed due to requirement 'sizeof(nlohmann::basic_json<std::map, std
::vector, std::string, bool, long long, unsigned long long, double, std::allocator, nlohmann::adl_serializer>) == 0' "could not find to_json() method in T's namespace"
        static_assert(sizeof(BasicJsonType) == 0,
        ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~
~/.conan/data/nlohmann_json/3.1.1/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/nlohmann/detail/conversions/to_json.hpp:328:16: note: in instantiation of function template specialization 'nlohmann::detail::to_json_fn
::call<nlohmann::basic_json<>, AER::Vector<std::complex<double>> &>' requested here
        return call(j, std::forward<T>(val), priority_tag<1> {});
               ^
~/.conan/data/nlohmann_json/3.1.1/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/nlohmann/adl_serializer.hpp:42:9: note: in instantiation of function template specialization 'nlohmann::detail::to_json_fn::operator()<n
lohmann::basic_json<>, AER::Vector<std::complex<double>> &>' requested here
        ::nlohmann::to_json(j, std::forward<ValueType>(val));
        ^
~/.conan/data/nlohmann_json/3.1.1/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/nlohmann/json.hpp:1241:28: note: in instantiation of function template specialization 'nlohmann::adl_serializer<AER::Vector<std::complex
<double>>>::to_json<nlohmann::basic_json<>, AER::Vector<std::complex<double>> &>' requested here
        JSONSerializer<U>::to_json(*this, std::forward<CompatibleType>(val));
                           ^
src/framework/results/data/subtypes/single_data.hpp:73:19: note: in instantiation of function template specialization 'nlohmann::basic_json<>::basic_json<AER::Vector<std::co
mplex<double>> &, AER::Vector<std::complex<double>>, 0>' requested here
  json_t jsdata = value();
                  ^
src/framework/results/data/subtypes/data_map.hpp:212:40: note: in instantiation of member function 'AER::SingleData<AER::Vector<std::complex<double>>>::to_json' requested he
re
      jsdata[pair.first] = pair.second.to_json();
                                       ^
src/framework/results/data/mixins/data_cvector.hpp:62:46: note: in instantiation of member function 'AER::DataMap<AER::SingleData, AER::Vector<std::complex<double>>, 1>::add
_to_json' requested here
  DataMap<SingleData, Vector<complex_t>, 1>::add_to_json(result);
```

I was not able to find which codes in the current main provide specialization.

**Comment by hhorii** [comment]
@doichanj 

> https://github.com/hhorii/qiskit-aer/blob/e7c66f66936c2a05d34c4b8251cad5df4812ae4b/src/simulators/statevector/statevector_state.hpp#L512-L516 there are snapshots in comment (also in other State classes) is JSON used for snapshots?

I think `json_chop_threshold_` is used in save instructions.

**Review by doichanj — APPROVED** [review]

**Review by jakelishman — APPROVED** [review]
The requirement that we put in manual `to_json` definitions for `Aer::Vector<std:complex>` when we didn't need to before and we don't know why is worrying to me, but given that the implementation of that looks fine, I guess that's not a huge deal.

**Pull request merged**